text/0000-cargo-report-future-incompat.md

```diff
@@ -0,0 +1,720 @@
+    for a crate somewhere in the `build/` directory, so that it can
+    emit the same report without having to rebuild the crate on subsequent
+    rebuilds of the root crate.
+
+  * Cargo is responsible for suggesting ways to address the problem to
+    the user. The specific tactics for constructing such suggestions
+    are not mandated by this RFC, but some ideas are presented as
+    [Future possibiilties][future-possibilities].
+
+## Implmentation strategy: Leverage JSON error-format
+
+The cleanest way to implement the above division of responsbilities
+without perturbing *non-cargo* uses of `rustc` is probably to make two changes:
+
+ * Cargo should *always* use `--error-format=json` for its `rustc`
+   invocations, and
+
+ * `rustc` should treat `--error-format=json` as signal that it should
+   emit a future-incompatibility summary report for the crate.
+
+It is relatively easy to extend the JSON output of `rustc` to include
+a new record of any future-incompatibility lints that were triggered
+when compiling a given crate.
+
+ * However, this RFC does not *dictate* this choice of implementation
+   strategy. (Other options include using some environment variable to
+   opt-in to a change to `rustc`'s output, or having `rustc` emit
+   future-incompatibility metadata to the filesystem.)
+
+Some (but not all) future-incompatibility lints will have a concrete
+schedule established for when they are meant to become hard errors.
+(This RFC does not specify the details about how such schedules are
+established or what constraints they will have to meet; it just posits
+that they *will* be established by some means.) The metadata for every
+future-incompatibility lint should include the anticipated version of
+Rust, if known, where it will become a hard error.
+
+ * This adds motivation for using JSON formatted diagnostics: JSON
+   records are more readily extensible, and thus can support adding
+   this sort of feedback in a robust fashion.
+
+Since `cargo` is expected to continue to emit the report even when the
+upstream dependencies are not rebuilt, Cargo will store the
+future-incompatibility status for each crate somewhere in the `build/`
+directory on the file-system. (This is analogous to how incremental
+compilation caches diagnostic output, so that future runs will still
+emit the same report even when we do not recompile the same input.)
+
+
+## Policy issues
+
+We probably do not want to blindly convert every lint that is
+currently categorized as `C-future-incompatility` on the Rust repo to
+use this system. For example, there may be cases where we would be better
+served to use the edition mechanism to introduce the hard-error, but leave
+the functionality unchanged for older Rust editions. In any case, that
+
+is a policy matter for the relevant teams, and is out of scope for
+this RFC. I just want to make it clear that the mechanism suggested
+here may not be appropriate for every single lint currently
```

Comment:
I've checked the list of C-future-compatibility issues and I did not see a single one which, in my view, should be addressed with an edition. Until the whole language team decides otherwise, all of the C-future-compatibility issues are bugs and should be fixed. Since the policy matter is out of scope for this RFC I would prefer for this section to be removed or at least moved out "Reference-level explanation".

Reply:
Revised text in 571478ea68c6e373f1ebae07aeede20b54999b99 to try to make the intent of these sentences clearer.

I cannot tell if @Centril 's complaint extends even to the original final sentence
> ... the mechanism suggested here may not be appropriate for every single lint currently categorized as `C-future-incompatility`

I am going to assume that bit is not contentious.